bears/jinja2/Jinja2Bear.py

```diff
@@ -0,0 +1,271 @@
+import re
+
+from coalib.bears.LocalBear import LocalBear
+from coalib.results.Result import Result, RESULT_SEVERITY
+from coalib.results.Diff import Diff
+
+
+def generate_spacing_diff(file, filename, line, line_number,
+                          match_object, required_spacing):
+    '''
+    Generate a diff for incorrectly spaced control or variable tags.
+
+    :param match_object:     the match object generated by a match on the
+                             regular expression control_start_regex,
+                             control_end_regex or variable_regex in the
+                             run method.
+    :param required_spacing: the number of spaces expected after the open
+                             delimiter and before the close delimiter
+    '''
+    diff = Diff(file)
+
+    content_before = line[:match_object.start("open")]
+    content_after = line[match_object.end("close"):]
+
+    spacing = " " * required_spacing
+    replacement = (
+        "{before}{open}{spacing}{content}{spacing}{close}{after}".format(
+            before=content_before,
+            spacing=spacing,
+            after=content_after,
+            content=match_object.group("content").strip(),
+            open=match_object.group("open"),
+            close=match_object.group("close")))
+    diff.change_line(
+        line_number,
+        line,
+        replacement)
+    return {filename: diff}
+
+
+def generate_label_diff(file, filename, line, line_number,
+                        match_object, expected_label):
+    '''
+    Generates a diff for a missing or wrong control loop end label.
+    Missing labels will be added, wrong ones replaced, content after the
+    label will be left untouched.
+
+
+    :param match_object:   the match object generated by a match on the regular
+                           expression control_end_regex in the run method
+    :param expected_label: The expected label for that control block.
+
+    '''
+    diff = Diff(file)
+
+    content_before = line[:match_object.end("close")]
+
+    suffix_start = match_object.end("close")
+    # if a label is there, we cut it out, by starting the kept suffix after it
+    if match_object.group("label") is not None:
+        suffix_start = match_object.end("label")
+
+    content_after = line[suffix_start:]
+
+    replacement = "{before}{label}{after}".format(
+        before=content_before,
+        after=content_after,
+        label=expected_label
+    )
+    diff.change_line(
+        line_number,
+        line,
+        replacement)
+    return {filename: diff}
+
+
+def is_spacing_valid(string, required_spacing):
+    '''
+    Returns True iff `string` has exactly <required_spacing>
+    spaces at the start and at the end
+    '''
+    leading_spaces = len(string) - len(string.lstrip())
```

Comment:
Trying to do complexity analysis here:
Assume length of string is L. Assume number of whitespaces is W.
Your current code does L + W + (L-W) = (for first length) + (for the lstrip) + (for the second length)
This can ideally be done in just W by a for loop:
    
    count = 0
    for s in string:
        count += 1
        if s != ' ':
            break

OR:

    for (i, s) in enumerate(string):
        if s != " ":
            count = i
            break

If you're interested, a quick benchmark would help in understanding this, otherwise it's not that important.